lib/AST/NameLookup.cpp

```diff
@@ -2847,15 +2846,18 @@ CollectedOpaqueReprs swift::collectOpaqueReturnTypeReprs(TypeRepr *r, ASTContext
       
       if (auto existential = dyn_cast<ExistentialTypeRepr>(repr)) {
         return Action::SkipChildren();
-      } else if (auto compositionRepr = dyn_cast<CompositionTypeRepr>(repr)) {
-        if (!compositionRepr->isTypeReprAny())
-          Reprs.push_back(compositionRepr);
+      } else if (auto composition = dyn_cast<CompositionTypeRepr>(repr)) {
+        if (!composition->isTypeReprAny())
+          Reprs.push_back(composition);
         return Action::SkipChildren();
       } else if (auto generic = dyn_cast<GenericIdentTypeRepr>(repr)) {
-        return Action::Continue();
-      } else if (auto declRefTR = dyn_cast<DeclRefTypeRepr>(repr)) {
-        if (declRefTR->isProtocol(dc))
-          Reprs.push_back(declRefTR);
+        if (generic->argsAreProtocols(dc))
+           return Action::Continue();
+        Reprs.push_back(generic);
```

Comment:
I think we need to gate this particular line behind `generic->isProtocol(dc)`, right? I think we also want to do this before determining whether the arguments contain protocols, otherwise for something like `Collection<P>`, the opaque type collector will only collect `P` and not `Collection` under `ImplicitSome`.

Reply:
ah you're right!